subsys/bluetooth/host/hci_core.c

```diff
@@ -3778,10 +3788,30 @@ int bt_disable(void)
 
 	bt_monitor_send(BT_MONITOR_CLOSE_INDEX, NULL, 0);
 
+#if defined(CONFIG_BT_PRIVACY)
+	k_work_cancel_delayable(&bt_dev.rpa_update);
+#endif
+
 #if defined(CONFIG_BT_PER_ADV_SYNC)
 	bt_periodic_sync_disable();
 #endif /* CONFIG_BT_PER_ADV_SYNC */
 
+#if defined(CONFIG_BT_CENTRAL) || defined(CONFIG_BT_OBSERVER)
+	bt_le_scan_uninit();
+#endif
+
+#if defined(CONFIG_BT_CONN)
+	err = bt_conn_auth_cb_register(NULL);
+	if (err) {
+		return err;
+	}
+
+	bt_pub_key_callback_clear();
+	bt_conn_cleanup_all_conns();
+	disconnected_handles_reset();
+	bt_conn_cb_unregister();
```

Comment:
Please add a sentence to `bt_conn_cb_register` API doc saying that the callbacks registered trough that function are automatically cleared by `bt_disable` after all connections are destroyed and all relevant callbacks have been invoked.

RFC: I suggest we don't clear the conn_cb listeners as part of `bt_disable`. It feels cleaner to make the listener registration independent. To further make it independent, we can explicitly allow registering before `bt_enable`. We can provide a `bt_conn_cb_unregister(struct bt_conn_cb *)` API.

Reply:
Agree with the RFC.

Reply:
Same with the scan_cbs and pub_key_cb_slist, do you reckon?

Reply:
I don't really agree with the RFC if the role of bt_disable is to perform a soft-reset of the whole bt subsystem. 
In that case, some state would linger on (in the form of those callbacks being registered). It also makes the mental model of what's happening clearer and forces the app users to not depend on said state.

I don't think a lot of users are currently calling `bt_disable` (we'd have more bug reports), so the few that do it can do the extra work of re-registering any callbacks they depend on.

Reply:
> Same with the scan_cbs and pub_key_cb_slist, do you reckon?

`scan_cbs` is the same, yes.

`pub_key_cb_slist` actually contains callbacks as opposed to event listeners. The callback entries are waiting for some operation to complete, and must all be invoked/completed exactly once. You can see in `bt_hci_evt_le_pkey_complete` that the operation can fail and the callback will then invoked with NULL. Instead of clearing this list manually, the operations that the callbacks are waiting on should be aborted and failed.